fl4health/clients/moon_client.py

```diff
@@ -0,0 +1,107 @@
+
+import torch
+from flwr.common.typing import Config, NDArrays
+
+from fl4health.checkpointing.checkpointer import TorchCheckpointer
+from fl4health.clients.basic_client import BasicClient
+from fl4health.model_bases.moon_base import MoonModel
+from fl4health.utils.losses import Losses, LossMeterType
+from fl4health.utils.metrics import Metric, MetricMeterType
+
+
+class MoonClient(BasicClient):
+    """
+    This client implements the MOON algorithm from Model-Contrastive Federated Learning. The key idea of MOON
+    is to utilize the similarity between model representations to correct the local training of individual parties,
+    i.e., conducting contrastive learning in model-level.
+    """
+
+    def __init__(
+        self,
+        data_path: Path,
+        metrics: Sequence[Metric],
+        device: torch.device,
+        loss_meter_type: LossMeterType = LossMeterType.AVERAGE,
+        metric_meter_type: MetricMeterType = MetricMeterType.AVERAGE,
+        checkpointer: Optional[TorchCheckpointer] = None,
+    ) -> None:
+        super().__init__(
+            data_path=data_path,
+            metrics=metrics,
+            device=device,
+            loss_meter_type=loss_meter_type,
+            metric_meter_type=metric_meter_type,
+            checkpointer=checkpointer,
+        )
+        self.initial_tensors: List[torch.Tensor]
+        self.contrastive_weight: float = 10
+        self.temprature: float = 0.5
+        self.current_loss: float
+        self.cos_sim = torch.nn.CosineSimilarity(dim=-1)
+        self.ce_criterion = torch.nn.CrossEntropyLoss().to(self.device)
+        self.len_old_models_buffer: int = 1
+        self.old_models_list: list[MoonModel] = []
+        self.global_model: MoonModel
+        self.features: torch.Tensor
+        self.old_features_list: list[torch.Tensor] = []
+        self.global_features: torch.Tensor
+
+    def predict(self, input: torch.Tensor) -> Dict[str, torch.Tensor]:
+        preds, self.features, _ = self.model(input)
+        self.features = self.features.view(len(self.features), -1)
+        self.old_features_list = []
+        for old_model in self.old_models_list:
+            _, old_features, _ = old_model(input)
+            old_features = old_features.view(len(old_features), -1).detach()
+            self.old_features_list.append(old_features)
+        _, self.global_features, _ = self.global_model(input)
+        self.global_features = self.global_features.view(len(self.global_features), -1).detach()
+        if isinstance(preds, dict):
+            return preds
+        elif isinstance(preds, torch.Tensor):
+            return {"prediction": preds}
+        else:
+            raise ValueError("Model forward did not return a tensor or dictionary or tensors")
+
+    def get_contrastive_loss(self) -> torch.Tensor:
+        assert len(self.features) == len(self.global_features)
+        posi = self.cos_sim(self.features, self.global_features)
```

Comment:
It would be good to add comments to clarify how the contrastive loss is computed. I see you are stacking the predictions of global model and subsequently old models column wise. Then setting labels to 0, indicating the similarity for positive pairs should be the highest when compared to the negative similarities in other columns. May not be clear to people not familiar with how MOON computes the contrastive loss.